### ISSUE TITLE
4.03.0

### ISSUE THREAD

**Issue by Drup** [OCamlPro/ocp-build#1]
Can we have a fix for current trunk ? I would like to use ocp-indent and ocp-browser.

cc @lefessan @AltGr

**Comment by lefessan** [comment]
Does this one work ?

https://github.com/ocaml/opam-repository/pull/3703

**Comment by Drup** [comment]
It seems to. Thanks.

**Comment by Drup** [comment]
Actually, it's broken again latest trunk now.

**Comment by francoisthire** [comment]
I have an issue trying to instal ocp-build : 

Undefined reference to caml_is_printable in the file simpleConfig

**Comment by ohad** [comment]
I have the same issue with 4.03.0

**Comment by ohad** [comment]
OK, here's one potential fix:

Apply the patch:

```
--- ../ocp-build.1.99.9-beta-original/src/ocplib/system/simpleConfig.ml4    2015-03-11 16:03:39.000000000 +0000
+++ ../ocp-build.1.99.9-beta-fix/src/ocplib/system/simpleConfig.ml4 2015-12-22 10:07:14.940363954 +0000
@@ -326,45 +326,6 @@

 let exit_exn = Exit

-
-let unsafe_get = String.unsafe_get
-external is_printable : char -> bool = "caml_is_printable"
-let unsafe_set = String.unsafe_set
-
-let escaped s =
-  let n = ref 0 in
-  for i = 0 to String.length s - 1 do
-    n :=
-      !n +
-        (match unsafe_get s i with
-           '"' | '\\' -> 2
-         | '\n' | '\t' -> 1
-         | c -> if is_printable c then 1 else 4)
-  done;
-  if !n = String.length s then s
-  else
-    let s' = String.create !n in
-    n := 0;
-    for i = 0 to String.length s - 1 do
-      begin match unsafe_get s i with
-        '"' | '\\' as c -> unsafe_set s' !n '\\'; incr n; unsafe_set s' !n c
-      | '\n' | '\t' as c -> unsafe_set s' !n c
-      | c ->
-          if is_printable c then unsafe_set s' !n c
-          else
-            let a = int_of_char c in
-            unsafe_set s' !n '\\';
-            incr n;
-            unsafe_set s' !n (char_of_int (48 + a / 100));
-            incr n;
-            unsafe_set s' !n (char_of_int (48 + a / 10 mod 10));
-            incr n;
-            unsafe_set s' !n (char_of_int (48 + a mod 10))
-      end;
-      incr n
-    done;
-    s'
-
 let safe_string s =
   if s = "" then "\"\""
   else
@@ -384,7 +345,7 @@
             s
           else raise exit_exn
     with
-      _ -> Printf.sprintf "\"%s\"" (escaped s)
+      _ -> Printf.sprintf "\"%s\"" (String.escaped s)

 let with_help = ref false
```

what it does is replace the `escaped` function with the [newly added](https://github.com/ocaml/ocaml/commit/e60a2db81c19bfa391107b921841c0abec30c458) String.escaped function.

This fix might mean some dependencies are added to the package, if this is to be incorporated to trunk.

I'm a complete Opam newbie, so here's what I did to apply it

```
$ opam source --dir <path for modified source> ocp-build
```

Now apply the patch to `<path>/src/ocplib/system/simpleConfig.ml4`:

Then:

```
$ opam pin add ocp-build <path>
```

and that should recompile everything appropriately.

Lots of kudos to edwin from #ocaml on Freenode who helped me.

**Comment by hcarty** [comment]
Failure with the 4.03.0+beta1 OCaml release:

```
=-=- Processing actions -=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=
[ERROR] The compilation of ocp-build failed at "make".
Processing  1/2: [ocp-build: sh]
#=== ERROR while installing ocp-build.1.99.13-beta ============================#
# opam-version 1.2.2
# os           linux
# command      make
# path         /home/vagrant/.opam/4.03.0+beta1/build/ocp-build.1.99.13-beta
# compiler     4.03.0+beta1
# exit-code    2
# env-file     /home/vagrant/.opam/4.03.0+beta1/build/ocp-build.1.99.13-beta/ocp-build-24436-6d07ae.env
# stdout-file  /home/vagrant/.opam/4.03.0+beta1/build/ocp-build.1.99.13-beta/ocp-build-24436-6d07ae.out
# stderr-file  /home/vagrant/.opam/4.03.0+beta1/build/ocp-build.1.99.13-beta/ocp-build-24436-6d07ae.err
### stdout ###
# [...]
# ocamlopt.opt -c -o ocplib-lang/stringSet.cmx -I ocplib-compat/string-compat -I ocplib-debug -I ocplib-lang -I ocplib-system -I ocplib-unix -I ocp-build ocplib-lang/stringSet.ml
# ocamlc.opt -c -o ocplib-lang/toposort.cmi -I ocplib-compat/string-compat -I ocplib-debug -I ocplib-lang -I ocplib-system -I ocplib-unix -I ocp-build ocplib-lang/toposort.mli
# ocamlopt.opt -c -o ocplib-lang/toposort.cmx -I ocplib-compat/string-compat -I ocplib-debug -I ocplib-lang -I ocplib-system -I ocplib-unix -I ocp-build ocplib-lang/toposort.ml
# ocamlc.opt -c -o ocplib-lang/linearToposort.cmi -I ocplib-compat/string-compat -I ocplib-debug -I ocplib-lang -I ocplib-system -I ocplib-unix -I ocp-build ocplib-lang/linearToposort.mli
# ocamlopt.opt -c -o ocplib-lang/linearToposort.cmx -I ocplib-compat/string-compat -I ocplib-debug -I ocplib-lang -I ocplib-system -I ocplib-unix -I ocp-build ocplib-lang/linearToposort.ml
# ocamlc.opt -c -o ocplib-lang/ocamllexer.cmi -I ocplib-compat/string-compat -I ocplib-debug -I ocplib-lang -I ocplib-system -I ocplib-unix -I ocp-build ocplib-lang/ocamllexer.mli
# ocamlopt.opt -c -o ocplib-lang/ocamllexer.cmx -I ocplib-compat/string-compat -I ocplib-debug -I ocplib-lang -I ocplib-system -I ocplib-unix -I ocp-build ocplib-lang/ocamllexer.ml
# ocamlc.opt -c -o ocplib-lang/trie.cmi -I ocplib-compat/string-compat -I ocplib-debug -I ocplib-lang -I ocplib-system -I ocplib-unix -I ocp-build ocplib-lang/trie.mli
# ocamlopt.opt -c -o ocplib-lang/trie.cmx -I ocplib-compat/string-compat -I ocplib-debug -I ocplib-lang -I ocplib-system -I ocplib-unix -I ocp-build ocplib-lang/trie.ml
# make[1]: Leaving directory `/home/vagrant/.opam/4.03.0+beta1/build/ocp-build.1.99.13-beta'
### stderr ###
# [...]
# Warning 3: deprecated: StringCompat.String.lowercase
# Use String.lowercase_ascii instead.
# File "ocplib-lang/trie.ml", line 1:
# Error: The implementation ocplib-lang/trie.ml
#        does not match the interface ocplib-lang/trie.cmi:
#        ...
#        At position module Make(M) : <here>
#        The value `union' is required but not provided
# make[1]: *** [ocplib-lang/trie.cmx] Error 2
# make: *** [boot/ocp-build.asm] Error 2
```

**Comment by hcarty** [comment]
A slightly updated patch.  Edit: this adapts to OCaml 4.03's new integer and float parsing (https://github.com/ocaml/ocaml/commit/9eca3e45191ef52ce519e2d5bb4688da62bf7496).  It builds but is otherwise untested.

```
diff --git a/ocp-pp/ocpp.ml b/ocp-pp/ocpp.ml
index 2c00d6a..9434525 100644
--- a/ocp-pp/ocpp.ml
+++ b/ocp-pp/ocpp.ml
@@ -260,15 +260,12 @@ let name_of_token token =
 *)

   | LIDENT string -> Printf.sprintf "LIDENT(%s)" string
-  | NATIVEINT(nativeint ) -> Printf.sprintf "NATIVEINT(%nd)" nativeint
   | OPTLABEL(string) -> Printf.sprintf "OPTLABEL(%s)" string
-  | INT int -> Printf.sprintf "INT(%d)" int
-  | INT32(int32) -> Printf.sprintf "INT32(%ld)" int32
-  | INT64(int64) -> Printf.sprintf "INT64(%Ld)" int64
+  | INT (int, s) -> Printf.sprintf "INT(%s,%s)" int (match s with None -> "None" | Some c -> Printf.sprintf "Some %c" c)
   | LABEL(string) -> Printf.sprintf "LABEL(%s)" string
   | UIDENT string -> Printf.sprintf  "UIDENT(%s)" string
   | CHAR char -> Printf.sprintf "CHAR(%S)" (String.make 1 char)
-  | FLOAT float -> Printf.sprintf "FLOAT(%s)" float
+  | FLOAT (float, s) -> Printf.sprintf "FLOAT(%s,%s)" float (match s with None -> "None" | Some c -> Printf.sprintf "Some %c" c)
   | PREFIXOP(string) -> Printf.sprintf "PREFIXOP(%s)" string
   | INFIXOP0(op) -> Printf.sprintf "INFIXOP0(%s)" op
   | INFIXOP1(op) -> Printf.sprintf "INFIXOP1(%s)" op
@@ -382,15 +379,12 @@ let string_of_token token =
       | BIGINT s -> Printf.sprintf "%sI" s *)

   | LIDENT string -> string
-  | NATIVEINT nativeint -> Printf.sprintf "%nd" nativeint
   | OPTLABEL(string) -> Printf.sprintf "?%s:" string
-  | INT int -> Printf.sprintf "%d" int
-  | INT32(int32) -> Printf.sprintf "%ld" int32
-  | INT64(int64) -> Printf.sprintf "%Ld" int64
+  | INT (int, s) -> Printf.sprintf "%s%s" int (match s with None -> "" | Some c -> Printf.sprintf "%c" c)
   | LABEL(string) -> Printf.sprintf "~%s:" string
   | UIDENT string -> Printf.sprintf  "%s" string
   | CHAR char -> Printf.sprintf "'%s'" (String.escaped (String.make 1 char))
-  | FLOAT float -> float
+  | FLOAT (float, s) -> Printf.sprintf "%s%s" float (match s with None -> "" | Some c -> Printf.sprintf "%c" c)
   | PREFIXOP(string) -> string
   | INFIXOP0(op) -> op
   | INFIXOP1(op) -> op
@@ -467,7 +461,7 @@ let parse_expr tokens lexbuf =
           [] -> P.EOF
         | token :: tail -> tokens := tail;
           match token with
-          | INT n -> P.INT n
+          | INT (n,None) -> P.INT (int_of_string n)
           | DOT -> P.DOT
           | TRUE -> P.TRUE
           | FALSE -> P.FALSE
@@ -906,7 +900,7 @@ and preprocess_directive lexer  directive =
       match args with
       | [ UIDENT name; STRING _ as token ] ->
         set_option name (Compat.get_STRING token)
-      | [ UIDENT name; INT s ] ->  set_option name (string_of_int s)
+      | [ UIDENT name; INT (n, None) ] ->  set_option name n
       | [ UIDENT name ] -> set_option name "1"
       | _ -> failwith "Error after #option: should be Option \"value\""
     end;
@@ -1224,7 +1218,7 @@ let _ =
   add_option "Warn" (fun name v -> Warnings.parse_options false v);
   add_option "Warn_error" (fun name v -> Warnings.parse_options true v);
 *)
-  add_macro "Has_ocpp" [ INT 1 ];
+  add_macro "Has_ocpp" [ INT ("1", None) ];
   register ()

 let lexbuf () = !current_lexbuf
diff --git a/ocplib-lang/trie.ml b/ocplib-lang/trie.ml
index b030909..6d59491 100644
--- a/ocplib-lang/trie.ml
+++ b/ocplib-lang/trie.ml
@@ -217,4 +217,8 @@ module Make (M : Map.S) = struct
     let bindings t =
       Printf.eprintf "TODO\n%!";
       assert false
+
+    let union t1 t2 =
+      Printf.eprintf "TODO\n%!";
+      assert false
 end
diff --git a/ocplib-system/simpleConfig.ml4 b/ocplib-system/simpleConfig.ml4
index 74a9a9b..4e1a9f5 100644
--- a/ocplib-system/simpleConfig.ml4
+++ b/ocplib-system/simpleConfig.ml4
@@ -336,45 +336,6 @@ let really_load filename sections =

 let exit_exn = Exit

-
-let unsafe_get = String.unsafe_get
-external is_printable : char -> bool = "caml_is_printable"
-let unsafe_set = Bytes.unsafe_set
-
-let escaped s =
-  let n = ref 0 in
-  for i = 0 to String.length s - 1 do
-    n :=
-      !n +
-        (match unsafe_get s i with
-           '"' | '\\' -> 2
-         | '\n' | '\t' -> 1
-         | c -> if is_printable c then 1 else 4)
-  done;
-  if !n = String.length s then s
-  else
-    let s' = Bytes.create !n in
-    n := 0;
-    for i = 0 to String.length s - 1 do
-      begin match unsafe_get s i with
-        '"' | '\\' as c -> unsafe_set s' !n '\\'; incr n; unsafe_set s' !n c
-      | '\n' | '\t' as c -> unsafe_set s' !n c
-      | c ->
-          if is_printable c then unsafe_set s' !n c
-          else
-            let a = int_of_char c in
-            unsafe_set s' !n '\\';
-            incr n;
-            unsafe_set s' !n (char_of_int (48 + a / 100));
-            incr n;
-            unsafe_set s' !n (char_of_int (48 + a / 10 mod 10));
-            incr n;
-            unsafe_set s' !n (char_of_int (48 + a mod 10))
-      end;
-      incr n
-    done;
-    Bytes.to_string s'
-
 let safe_string s =
   if s = "" then "\"\""
   else
@@ -394,7 +355,7 @@ let safe_string s =
             s
           else raise exit_exn
     with
-      _ -> Printf.sprintf "\"%s\"" (escaped s)
+      _ -> Printf.sprintf "\"%s\"" (String.escaped s)

 let with_help = ref false

```

**Comment by lefessan** [comment]
Thanks for the patch !
Commit af45fee4a189827b5c30ff8b2ef9fa6d0d8364fb should provide a complete solution (compatibility from 3.12.1 to 4.03.0). I will release 1.99.14-beta as soon as back from holidays.

**Comment by lefessan** [comment]
Fixed in 1.99.14-beta.
